src/Microsoft.Identity.Client/Telemetry.cs

```diff
@@ -42,11 +45,144 @@ public void RegisterReceiver(Receiver r)
 
         private static readonly Telemetry Singleton = new Telemetry();
 
-        private Telemetry(){}
+        internal Telemetry(){}  // This is an internal constructor to build isolated unit test instance
 
         public static Telemetry GetInstance()
         {
             return Singleton;
         }
+
+        public bool TelemetryOnFailureOnly { get; set; }
+
+        private ConcurrentDictionary<Tuple<string, string>, EventBase> EventsInProgress = new ConcurrentDictionary<Tuple<string, string>, EventBase>();
+
+        private ConcurrentDictionary<string, List<EventBase>> CompletedEvents = new ConcurrentDictionary<string, List<EventBase>>();
+
+        internal string GenerateNewRequestId()
+        {
+            return Guid.NewGuid().ToString();
+        }
+
+        internal void StartEvent(string requestId, EventBase eventToStart)
+        {
+            if (requestId != null)
```

Comment:
What if we think it this way? "if there is no receiver defined, then it means the app dev does not need telemetry feature", this is a higher level business logic, and there can be many different ways to implement it. Right now, both Android Telemetry PR and the current PR implements that in the `Flush()`.

Here at this particular line, it is inside `StartEvent()`. The current check is not exactly a business logic. It is more like an implementation detail. Here the check tries to guard against a potentially empty `requestId`, before we are going to run some logic based on a hopefully not-null `requestId`. It used to help me during early development when things were not in order. Now it is not strictly necessary though, the Android PR does not even use this check. So if you prefer, I think we can simply remove this line. Let me know.